File: tang/src/test/java/com/microsoft/tang/TweetExample.java
package com.microsoft.tang;

import javax.inject.Inject;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import com.microsoft.tang.annotations.Name;
import com.microsoft.tang.annotations.NamedParameter;
import com.microsoft.tang.annotations.Parameter;
import com.microsoft.tang.implementation.ConfigurationBuilderImpl;
import com.microsoft.tang.implementation.InjectorImpl;


public class TweetExample {
  private static interface TweetFactory {
    public String getTweet();
  }

  private static interface SMS {
    public void sendSMS(String msg, long phoneNumber);
  }

  private static class MockTweetFactory implements TweetFactory {
    @Inject
    public MockTweetFactory() {
    }

    @Override
    public String getTweet() {
      return "@tw #bbq bbqftw!!! gopher://vuwed.wefd/bbqftw!";
    }
  }

  private static class MockSMS implements SMS {
    @Inject
    public MockSMS() {
    }

    @Override
    public void sendSMS(String msg, long phoneNumber) {
      if (phoneNumber != 867 - 5309) {
        throw new IllegalArgumentException("Unknown recipient");
      }
      // success!
    }
  }

  private static class Tweeter {
    final TweetFactory tw;
    final SMS sms;
    final long phoneNumber;

    @NamedParameter()
    class PhoneNumber implements Name<Long> {}
    @Inject
    public Tweeter(TweetFactory tw, SMS sms,
        @Parameter(PhoneNumber.class) long phoneNumber) {
      this.tw = tw;
      this.sms = sms;
      this.phoneNumber = phoneNumber;
    }

    public void sendMessage() {
      sms.sendSMS(tw.getTweet(), phoneNumber);
    }
  }

  @BeforeClass
  public static void setUpBeforeClass() throws Exception {
  }

  @AfterClass
  public static void tearDownAfterClass() throws Exception {
  }

  Tang tang;
  @Before
  public void setUp() throws Exception {
    tang = Tang.Factory.getTang();
  }


  @After
  public void tearDown() throws Exception {
  }

  @Test
  public void test() throws Exception {
    ConfigurationBuilderImpl t = (ConfigurationBuilderImpl)tang.newConfigurationBuilder();
    t.register(Tweeter.class);
    t.bindImplementation(TweetFactory.class, MockTweetFactory.class);
    t.bindImplementation(SMS.class, MockSMS.class);
    t.bindNamedParameter(Tweeter.PhoneNumber.class, new Long(867 - 5309).toString());
    Tweeter tw = (Tweeter) new InjectorImpl(t.build()).getInstance(Tweeter.class);
    tw.sendMessage();
  }

}

File: reef-tang/tang/src/main/cs/Examples/TweetExample.cs
/**
 * Copyright (C) 2014 Microsoft Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Com.Microsoft.Tang.Annotations;

namespace Com.Microsoft.Tang.Examples
{
    public interface SMS
    {
        void SendSMS(String msg, long phoneNumber);
    }

    public interface TweetFactory
    {
        String GetTweet();
    } 

    public class MockTweetFactory : TweetFactory 
    {
        [Inject]
        public MockTweetFactory() {
        }

        public String GetTweet() 
        {
            return "@tw #bbq bbqftw!!! gopher://vuwed.wefd/bbqftw!";
        }
    }

    public class MockSMS : SMS 
    {
        [Inject]
        public MockSMS() 
        {
        }

        public void SendSMS(String msg, long phoneNumber) 
        {
            if (phoneNumber != 8675309) 
            {
                throw new ArgumentException("Unknown recipient");
            }
            // success!
        }
    }

    public class Tweeter
    {
        TweetFactory tw;
        SMS sms;
        long phoneNumber;

        [NamedParameter("Phone number", "number", "1800")]
        class PhoneNumber : Name<long> { }
        [Inject]
        public Tweeter(TweetFactory tw, SMS sms, [Parameter(Value = typeof(PhoneNumber))] long phoneNumber)
        {
            this.tw = tw;
            this.sms = sms;
            this.phoneNumber = phoneNumber;
        }

        [Inject]
        public Tweeter([Parameter(Value = typeof(PhoneNumber))] long phoneNumber)
        {
            this.phoneNumber = phoneNumber;
        }

        public void sendMessage()
        {
            sms.SendSMS(tw.GetTweet(), phoneNumber);
        }
    }
}
